File: day15/ans.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        ans = [1]
        
        for i in range(len(nums) - 1):
            ans.append(nums[i] * ans[-1])
        
        idx = len(ans) - 1
        r = 1
        for i, j, k in zip(reversed(ans), reversed(nums), range(idx, -1, -1)):
            ans[k] = i*r
            r *= j
            
        return ans

File: day15/ans.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> forward = {1};
        vector<int> backward = {1};
        
        for(int i = 0, j = 0; i < nums.size() - 1; i++, j++){
           forward.push_back(nums[i] * forward[j]);
        }
        
        for(int i = nums.size() - 1, j = 0; i > 0; i--, j++){
            backward.push_back(nums[i] * backward[j]);
        }
        
        vector<int> product_array;
        int i = 0;
        int j = backward.size() - 1;
        
        while(i < forward.size() && j > -1) {
            product_array.push_back(forward[i] * backward[j]);
            i++;
            j--;
        }
        
        return product_array;
    }
};




